src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs

```diff
@@ -67,31 +68,32 @@ internal sealed partial class SocketConnection : TransportConnection
 
             var pair = DuplexPipe.CreateConnectionPair(inputOptions, outputOptions);
 
-            // Set the transport and connection id
-            Transport = _originalTransport = pair.Transport;
+            _originalTransport = pair.Transport;
             Application = pair.Application;
 
+            Transport = new SocketDuplexPipe(this);
+
             InitiaizeFeatures();
         }
 
+        public IDuplexPipe InnerTransport => _originalTransport;
+
         public PipeWriter Input => Application.Output;
 
         public PipeReader Output => Application.Input;
 
         public override MemoryPool<byte> MemoryPool { get; }
 
-        public void Start()
+        private void EnsureStarted()
         {
-            try
+            if (_connectionStarted == 1 || Interlocked.CompareExchange(ref _connectionStarted, 1, 0) == 1)
```

Comment:
For the "fast-check" `if (_connectionStarted == 1` should this be a volatile read?

Or from different view: `EnsureStarted` is only called in `DisposeAsync` so is it worth to have a fast-check? Just the compare & exchange should be enough?

Reply:
No, it's called in more than 2 places. The fast check is in the hot path (every read/write/flush)

Reply:
Ah, it's a partial class. Thanks.

But is still safe without a volatile read? Now I believe it's safe, as in the case of `_connectionStarted = 0` the CompareExchange is hit which has the proper memory-fences.